front/pages/api/data_sources/[user]/[name]/search.ts

```diff
@@ -0,0 +1,186 @@
+
+type TimestampFilter = {
+  gt?: number;
+  lt?: number;
+};
+
+type SearchFilter = {
+  tags?: TagsFilter;
+  timestamp?: TimestampFilter;
+};
+
+type DatasourceSearchQuery = {
+  query: string;
+  top_k: number;
+  filter?: SearchFilter;
+  full_text: boolean;
+};
+
+const search_query_schema: JSONSchemaType<DatasourceSearchQuery> = {
+  type: "object",
+  properties: {
+    query: { type: "string" },
+    top_k: { type: "number" },
+    full_text: { type: "boolean" },
+    filter: {
+      type: "object",
+      nullable: true,
+      properties: {
+        tags: {
+          type: "object",
+          nullable: true,
+          properties: {
+            is_in: { type: "array", items: { type: "string" }, nullable: true },
+            is_not: {
+              type: "array",
+              items: { type: "string" },
+              nullable: true,
+            },
+          },
+        },
+        timestamp: {
+          type: "object",
+          nullable: true,
+          properties: {
+            gt: { type: "number", nullable: true },
+            lt: { type: "number", nullable: true },
+          },
+        },
+      },
+    },
+  },
+  required: ["query", "top_k", "full_text"],
+};
+
+const { DUST_API } = process.env;
+
+type DatasourceSearchResponseBody = {
+  documents: Array<Document>;
+};
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<DatasourceSearchResponseBody | ApiError>
+) {
+  const session = await getServerSession(req, res, authOptions);
+
+  let user = await User.findOne({
+    where: {
+      username: req.query.user,
+    },
+  });
+
+  if (!user) {
+    res.status(404).end();
+    return;
+  }
+
+  const readOnly = !(
+    session && session.provider.id.toString() === user.githubId
+  );
+
+  let dataSource = await DataSource.findOne({
+    where: readOnly
+      ? {
+          userId: user.id,
+          name: req.query.name,
+          visibility: {
+            [Op.or]: ["public"],
+          },
+        }
+      : {
+          userId: user.id,
+          name: req.query.name,
+        },
+    attributes: [
+      "id",
+      "name",
+      "description",
+      "visibility",
+      "config",
+      "dustAPIProjectId",
+      "updatedAt",
+    ],
+  });
+
+  if (!dataSource) {
+    res.status(404).end();
+    return;
+  }
+
+  switch (req.method) {
+    case "POST":
+      let search_query: DatasourceSearchQuery | null = null;
+      try {
+        search_query = parse_payload(search_query_schema, req.body);
+        console.log("successfull parse", JSON.stringify(search_query));
+      } catch (e) {
+        if (e instanceof RequestParseError) {
+          res.status(400).json({
+            error: {
+              type: "invalid_request_error",
+              message: e.message,
+            },
+          });
+          return;
+        }
+      }
+
+      let [providers] = await Promise.all([
+        Provider.findAll({
+          where: {
+            userId: user.id,
```

Comment:
The credentials should be the ones of the logged in user not the DataSource’s otherwise I could spend money on the owner of a public DataSource